File: 6/solution.py
def tests():
    return [
        ("""COM)B
B)C
C)D
D)E
E)F
B)G
G)H
D)I
E)J
J)K
K)L
K)YOU
I)SAN""", "54 4")
    ]


class Node:
    def __init__(self):
        self._parent = None


def solve(data: str) -> str:
    nodes = {}
    for line in data.splitlines():
        a, b = line.split(')')
        if a in nodes:
            na = nodes[a]
        else:
            na = Node()
            nodes[a] = na
        if b in nodes:
            nb = nodes[b]
        else:
            nb = Node()
            nodes[b] = nb
        nb._parent = na

    # Part 1
    # Brute force
    # From every node count back to root
    orbits = 0
    for n in nodes.values():
        cur = n
        while True:
            if not cur._parent:
                break
            orbits += 1
            cur = cur._parent

    # Part 2
    # walk back from both nodes until matching parent found
    start = nodes['YOU']._parent
    end = nodes['SAN']._parent
    walk_start = []
    cur = start
    while cur._parent:
        cur = cur._parent
        walk_start.append(cur)
    walk_end = []
    cur = end
    while cur._parent:
        cur = cur._parent
        walk_end.append(cur)

    for i, node in enumerate(walk_start, 1):
        if node in walk_end:
            # Found match
            node_index = walk_end.index(node)
            transfers = i + node_index + 1
            break

    return f"{orbits} {transfers}"

File: 6/solution.rs
pub mod question6 {
    use std::collections::{HashMap, HashSet};

    pub fn solve(data: Vec<String>) -> String {
        // Parse input
        let mut nodes: HashSet<_> = HashSet::new();
        let mut orbits: HashMap<_, _> = HashMap::new();
        for line in &data {
            let mut line_split = line.split(')');
            let n1 = line_split.next().unwrap();
            let n2 = line_split.next().unwrap();

            nodes.insert(n1);
            nodes.insert(n2);

            orbits.insert(n2, n1);
        }

        // Part 1: Count total orbits
        let mut total_orbits: i64 = 0;
        for node in &nodes {
            let mut cur = node;
            loop {
                if !orbits.contains_key(cur) {
                    break;
                }
                total_orbits += 1;
                cur = &orbits[cur];
            }
        }

        // Part 2: Find shortest path between YOU and SAN

        // Returns a vector of orbits from start mass to COM
        fn walk<'a>(start: &'a str, orbits: &'a HashMap<&str, &str>) -> Vec<&'a str> {
            let mut chain: Vec<_> = Vec::new();
            let mut cur = start;
            loop {
                if !orbits.contains_key(cur) {
                    break;
                }
                cur = &orbits[cur];
                chain.push(cur);
            }

            return chain;
        }
        let w1 = walk("YOU", &orbits);
        let w2 = walk("SAN", &orbits);

        let mut walk_dist = 0;
        for mass in &w1 {
            if w2.contains(mass) {
                let w1_dist = w1.iter().position(|&x| &x == mass).unwrap();
                let w2_dist = w2.iter().position(|&x| &x == mass).unwrap();
                walk_dist = w1_dist + w2_dist;
                break;
            }
        }

        // Return solutions
        return format!("{} {}", total_orbits, walk_dist);
    }
}
